plutus-core/plutus-core/src/Universe/Core.hs

```diff
@@ -0,0 +1,756 @@
+is of kind star whenever we expect to parse/decode a constant of that type next. This is one reason
+why we have all these 'Typeable' constraints lying around.
+
+Type-wise this is enforced via 'bring' expecting a @uni (T a)@ with @a@ being of kind @Type@.
+I.e. there is no way one could parse a general type and then attempt to bring a constraint for
+that type in scope via 'bring' without first ensuring that the type is of kind star.
+
+Another reason for having the 'Typeable' constraints is that decoding a type application requires
+checking that the expected kind of an argument matches the kind of the provided argument due to us
+having an intrinsically-kinded reprensentation of types.
+-}
+
+{- Note [The TypeApp approach]
+There's an alternative approach to representing universes supporting partial application of
+type constructors. The idea is that we only "shield" those types that are of a non-star kind instead
+of applying @T@ to every single type in the universe. I.e. we specifically embed into @Type@ only
+(possibly partial) type applications and add one more constructor that allows for "running" such
+applications:
+
+    data TypeApp (a :: k)
+
+    data U (a :: Type) where
+        UProtoList  :: U (TypeApp [])
+        UInt        :: U Int
+        UApply      :: U (TypeApp f) -> U a -> U (TypeApp (f a))
+        URunTypeApp :: U (TypeApp a) -> U a
+
+This representation has the advantage of allowing for sharing the universes infrastructure between
+the monomorphic and polymorphic cases. I.e. we could use good old
+
+    data ValueOf uni a = ValueOf (uni a) a
+
+instead of having to use the slightly more awkward
+
+    data ValueOf uni a = ValueOf (uni (T a)) a
+
+One problem that this representation has is redundancy: there are two ways to represent the type
+of lists of integers (for some definition of @SomeTypeIn@):
+
+    SomeTypeIn (URunTypeApp (UProtoList `UApply` UInt))
+    SomeTypeIn (UProtoList `UApply` UInt)
+
+In practice it's not a big deal, we just need not to forget to strip the 'URunTypeApp' constructor
+off whenever that is important (for example, during type normalization).
+
+But a more serious problem with this representation is that we lose all the type safety discussed
+in Note [Decoding universes], which makes it rather easy to shoot oneself in the foot with
+@error "not supposed to happen"@ happening. I.e. we'd need to perform all the same checks
+in client code but without the type checker forcing us to do so like it does now, so that's a huge
+price to pay for some superficial syntactic nicety and hence we choose the safe approach,
+even though that required reworking all the infrastructure in a backwards-incompatible manner.
+-}
+
+-- See Note [Representing polymorphism].
+type T :: forall k. k -> Type
+data T a
+
+-- | Existential quantification as a data type.
+type Some :: forall a. (a -> Type) -> Type
+data Some f = forall x. Some !(f x)
```

Comment:
I specifically abstracted the whole thing out of Plutus as it's a very general machinery, so this is kinda like a library, so it might eventually be studied by people and explicit bangs help readability a lot compared to a sneaky pragma. So I'm not against `StrictData` in general and we could have it in this file as well, but I'd prefer to keep the explicit bangs.